Create RestClient bean from a RestHighLevelClient if such a bean has been contributed by the application.

nosan opened the pull request:
see gh-17464

bclozel commented on spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/elasticsearch/rest/RestClientAutoConfiguration.java:
```diff
@@ -92,4 +93,17 @@ public RestHighLevelClient restHighLevelClient(RestClientBuilder restClientBuild
+
+		@Bean
+		@ConditionalOnMissingBean
+		public RestClient restClient(RestHighLevelClient restHighLevelClient) {
```
There's already another bean method named for `RestClient` with the same name. Wouldn't that create issues with bean overriding?
Maybe we could change the existing bean method and inject `ObjectProvider<RestHighLevelClient>` and do everything there (return the client from the high level if it exists or create a new one from the builder). What fo you think (I've quickly looked at your PR, so I might haved missed something)?

nosan commented:
Thank you, @bclozel 

> There's already another bean method named for RestClient with the same name. Wouldn't that create issues with bean overriding?

Everything is ok, at least I added a test to check this.

> Maybe we could change the existing bean method and inject ObjectProvider<RestHighLevelClient>
 
We could not change the existing bean method as `RestHighLevelClient` and `RestClient` are present in different jars.
Use `ObjectProvider<RestHighLevelClient>` instead of `RestHighLevelClient` bean is a good idea 👍,
at least it is a guarantee that configuration will be applied only if there is a unique bean of `RestHighLevelClient`. (I added a test to check this as well)


PR has been updated.

snicoll commented:
Thanks again for the pr @nosan.

> Everything is ok, at least I added a test to check this.

It is ok indeed as they are conditions to avoid creating it. However `restClient` is quite generic and I think we should rename it. I've created #17725 

@nosan the first test in your PR has [an invalid assertion](https://github.com/spring-projects/spring-boot/pull/17488/files#diff-4010630ff478b9d04bdbb26d34a2765bR62). I fix it to assert that the rest client is the expected one and, of course, it failed since I've changed the assertion. The test failed as things were not created in the expected order.

Unrelated to your PR but there was a conceptual cycle as the inner class (processed first) required the `RestClientBuilder` that is provided by the auto-configuration class itself. I've polished the auto-configuration to avoid this cycle and provide an explicit order.

nosan commented:
Thank you so much, @snicoll 
> the first test in your PR has an invalid assertion. I fix it to assert that the rest client is the expected one and, of course, it failed since I've changed the assertion. The test failed as things were not created in the expected order.

 Sorry about that :confused:
